File: apps/go/chapter08/linked_list.go
// Package chapter08 第8章 リスト（連結リスト）
package chapter08

// --- 単方向連結リスト ---

type node struct {
	data int
	next *node
}

// LinkedList 単方向連結リスト
type LinkedList struct {
	head *node
	no   int
}

// NewLinkedList 新しい連結リストを生成する
func NewLinkedList() *LinkedList {
	return &LinkedList{}
}

// Len リストの要素数を返す
func (l *LinkedList) Len() int { return l.no }

// Contains リストにdataが含まれるかを確認する
func (l *LinkedList) Contains(data int) bool {
	ptr := l.head
	for ptr != nil {
		if ptr.data == data {
			return true
		}
		ptr = ptr.next
	}
	return false
}

// AddFirst 先頭にノードを挿入する
func (l *LinkedList) AddFirst(data int) {
	l.head = &node{data: data, next: l.head}
	l.no++
}

// AddLast 末尾にノードを挿入する
func (l *LinkedList) AddLast(data int) {
	n := &node{data: data}
	if l.head == nil {
		l.head = n
	} else {
		ptr := l.head
		for ptr.next != nil {
			ptr = ptr.next
		}
		ptr.next = n
	}
	l.no++
}

// RemoveFirst 先頭ノードを削除する
func (l *LinkedList) RemoveFirst() {
	if l.head != nil {
		l.head = l.head.next
		l.no--
	}
}

// RemoveLast 末尾ノードを削除する
func (l *LinkedList) RemoveLast() {
	if l.head == nil {
		return
	}
	if l.head.next == nil {
		l.head = nil
	} else {
		ptr := l.head
		for ptr.next != nil && ptr.next.next != nil {
			ptr = ptr.next
		}
		ptr.next = nil
	}
	l.no--
}

// Clear 全ノードを削除する
func (l *LinkedList) Clear() {
	l.head = nil
	l.no = 0
}

// --- 双方向連結リスト ---

type dnode struct {
	data int
	prev *dnode
	next *dnode
}

// DoublyLinkedList 双方向連結リスト
type DoublyLinkedList struct {
	head *dnode
	tail *dnode
	no   int
}

// NewDoublyLinkedList 新しい双方向連結リストを生成する
func NewDoublyLinkedList() *DoublyLinkedList {
	return &DoublyLinkedList{}
}

// Len リストの要素数を返す
func (l *DoublyLinkedList) Len() int { return l.no }

// Contains リストにdataが含まれるかを確認する
func (l *DoublyLinkedList) Contains(data int) bool {
	ptr := l.head
	for ptr != nil {
		if ptr.data == data {
			return true
		}
		ptr = ptr.next
	}
	return false
}

// AddFirst 先頭にノードを挿入する
func (l *DoublyLinkedList) AddFirst(data int) {
	n := &dnode{data: data, next: l.head}
	if l.head != nil {
		l.head.prev = n
	} else {
		l.tail = n
	}
	l.head = n
	l.no++
}

// AddLast 末尾にノードを挿入する
func (l *DoublyLinkedList) AddLast(data int) {
	n := &dnode{data: data, prev: l.tail}
	if l.tail != nil {
		l.tail.next = n
	} else {
		l.head = n
	}
	l.tail = n
	l.no++
}

// RemoveFirst 先頭ノードを削除する
func (l *DoublyLinkedList) RemoveFirst() {
	if l.head == nil {
		return
	}
	l.head = l.head.next
	if l.head != nil {
		l.head.prev = nil
	} else {
		l.tail = nil
	}
	l.no--
}

// Clear 全ノードを削除する
func (l *DoublyLinkedList) Clear() {
	l.head = nil
	l.tail = nil
	l.no = 0
}

// --- 配列カーソル版連結リスト ---

const nullIdx = -1

type arrayNode struct {
	data  int
	next  int
	dnext int // 削除済みリストの次インデックス
}

// ArrayLinkedList 配列カーソル版線形リスト
type ArrayLinkedList struct {
	n        []arrayNode
	head     int
	deleted  int
	max      int
	capacity int
	no       int
}

// NewArrayLinkedList 指定容量の配列カーソル版リストを生成する
func NewArrayLinkedList(capacity int) *ArrayLinkedList {
	nodes := make([]arrayNode, capacity)
	for i := range nodes {
		nodes[i] = arrayNode{next: nullIdx, dnext: nullIdx}
	}
	return &ArrayLinkedList{
		n:        nodes,
		head:     nullIdx,
		deleted:  nullIdx,
		max:      nullIdx,
		capacity: capacity,
	}
}

// Len リストの要素数を返す
func (al *ArrayLinkedList) Len() int { return al.no }

func (al *ArrayLinkedList) getInsertIndex() int {
	if al.deleted != nullIdx {
		rec := al.deleted
		al.deleted = al.n[rec].dnext
		return rec
	}
	if al.max+1 < al.capacity {
		al.max++
		return al.max
	}
	return nullIdx
}

// AddFirst 先頭にノードを挿入する
func (al *ArrayLinkedList) AddFirst(data int) {
	ptr := al.head
	rec := al.getInsertIndex()
	if rec == nullIdx {
		return
	}
	al.head = rec
	al.n[rec] = arrayNode{data: data, next: ptr}
	al.no++
}

// AddLast 末尾にノードを挿入する
func (al *ArrayLinkedList) AddLast(data int) {
	if al.head == nullIdx {
		al.AddFirst(data)
		return
	}
	ptr := al.head
	for al.n[ptr].next != nullIdx {
		ptr = al.n[ptr].next
	}
	rec := al.getInsertIndex()
	if rec == nullIdx {
		return
	}
	al.n[ptr].next = rec
	al.n[rec] = arrayNode{data: data, next: nullIdx}
	al.no++
}

// Search dataと等しいノードを探索してインデックスを返す（見つからなければ nullIdx）
func (al *ArrayLinkedList) Search(data int) int {
	ptr := al.head
	for ptr != nullIdx {
		if al.n[ptr].data == data {
			return ptr
		}
		ptr = al.n[ptr].next
	}
	return nullIdx
}

// RemoveFirst 先頭ノードを削除する
func (al *ArrayLinkedList) RemoveFirst() {
	if al.head == nullIdx {
		return
	}
	ptr := al.head
	al.head = al.n[ptr].next
	al.n[ptr].dnext = al.deleted
	al.deleted = ptr
	al.no--
}

File: apps/c/chapter08/linked_list.c
#include <stdlib.h>
#include "linked_list.h"

/* ===== 単方向連結リスト ===== */

LinkedList *ll_new(void) {
    LinkedList *lst = (LinkedList *)malloc(sizeof(LinkedList));
    lst->head = NULL;
    lst->no = 0;
    return lst;
}

void ll_free(LinkedList *lst) {
    ll_clear(lst);
    free(lst);
}

int ll_len(const LinkedList *lst) { return lst->no; }

int ll_get_head(const LinkedList *lst) {
    return lst->head ? lst->head->data : AL_NULL;
}

void ll_add_first(LinkedList *lst, int data) {
    LLNode *node = (LLNode *)malloc(sizeof(LLNode));
    node->data = data;
    node->next = lst->head;
    lst->head = node;
    lst->no++;
}

void ll_add_last(LinkedList *lst, int data) {
    LLNode *node = (LLNode *)malloc(sizeof(LLNode));
    node->data = data;
    node->next = NULL;
    if (lst->head == NULL) {
        lst->head = node;
    } else {
        LLNode *ptr = lst->head;
        while (ptr->next != NULL) ptr = ptr->next;
        ptr->next = node;
    }
    lst->no++;
}

void ll_remove_first(LinkedList *lst) {
    if (lst->head == NULL) return;
    LLNode *old = lst->head;
    lst->head = old->next;
    free(old);
    lst->no--;
}

void ll_remove_last(LinkedList *lst) {
    if (lst->head == NULL) return;
    if (lst->head->next == NULL) {
        free(lst->head);
        lst->head = NULL;
    } else {
        LLNode *ptr = lst->head;
        while (ptr->next != NULL && ptr->next->next != NULL)
            ptr = ptr->next;
        free(ptr->next);
        ptr->next = NULL;
    }
    lst->no--;
}

int ll_search(const LinkedList *lst, int data) {
    LLNode *ptr = lst->head;
    while (ptr != NULL) {
        if (ptr->data == data) return 1;
        ptr = ptr->next;
    }
    return 0;
}

void ll_clear(LinkedList *lst) {
    LLNode *ptr = lst->head;
    while (ptr != NULL) {
        LLNode *next = ptr->next;
        free(ptr);
        ptr = next;
    }
    lst->head = NULL;
    lst->no = 0;
}

/* ===== 双方向連結リスト（番兵ノード使用） ===== */

DList *dl_new(void) {
    DList *dl = (DList *)malloc(sizeof(DList));
    dl->head = (DLNode *)malloc(sizeof(DLNode));
    dl->head->data = 0;
    dl->head->prev = dl->head;
    dl->head->next = dl->head;
    dl->no = 0;
    return dl;
}

void dl_free(DList *dl) {
    dl_clear(dl);
    free(dl->head);
    free(dl);
}

int dl_len(const DList *dl) { return dl->no; }

void dl_add_first(DList *dl, int data) {
    DLNode *node = (DLNode *)malloc(sizeof(DLNode));
    node->data = data;
    node->prev = dl->head;
    node->next = dl->head->next;
    dl->head->next->prev = node;
    dl->head->next = node;
    dl->no++;
}

void dl_add_last(DList *dl, int data) {
    DLNode *node = (DLNode *)malloc(sizeof(DLNode));
    node->data = data;
    node->next = dl->head;
    node->prev = dl->head->prev;
    dl->head->prev->next = node;
    dl->head->prev = node;
    dl->no++;
}

void dl_remove_node(DList *dl, DLNode *node) {
    if (dl->no == 0) return;
    node->prev->next = node->next;
    node->next->prev = node->prev;
    free(node);
    dl->no--;
}

int dl_search(const DList *dl, int data) {
    DLNode *ptr = dl->head->next;
    while (ptr != dl->head) {
        if (ptr->data == data) return 1;
        ptr = ptr->next;
    }
    return 0;
}

void dl_clear(DList *dl) {
    DLNode *ptr = dl->head->next;
    while (ptr != dl->head) {
        DLNode *next = ptr->next;
        free(ptr);
        ptr = next;
    }
    dl->head->prev = dl->head;
    dl->head->next = dl->head;
    dl->no = 0;
}

/* ===== 配列カーソル版リスト ===== */

AList *al_new(int capacity) {
    AList *al = (AList *)malloc(sizeof(AList));
    al->head = AL_NULL;
    al->current = AL_NULL;
    al->max = AL_NULL;
    al->deleted = AL_NULL;
    al->capacity = capacity;
    al->n = (ANode *)malloc((size_t)capacity * sizeof(ANode));
    for (int i = 0; i < capacity; i++) {
        al->n[i].data = AL_NULL;
        al->n[i].next = AL_NULL;
        al->n[i].dnext = AL_NULL;
    }
    al->no = 0;
    return al;
}

void al_free(AList *al) {
    free(al->n);
    free(al);
}

int al_len(const AList *al) { return al->no; }

static int al_get_insert_index(AList *al) {
    if (al->deleted == AL_NULL) {
        if (al->max + 1 < al->capacity) {
            al->max++;
            return al->max;
        }
        return AL_NULL;
    } else {
        int rec = al->deleted;
        al->deleted = al->n[rec].dnext;
        return rec;
    }
}

void al_add_first(AList *al, int data) {
    int ptr = al->head;
    int rec = al_get_insert_index(al);
    if (rec == AL_NULL) return;
    al->head = al->current = rec;
    al->n[rec].data = data;
    al->n[rec].next = ptr;
    al->n[rec].dnext = AL_NULL;
    al->no++;
}

void al_add_last(AList *al, int data) {
    if (al->head == AL_NULL) {
        al_add_first(al, data);
        return;
    }
    int ptr = al->head;
    while (al->n[ptr].next != AL_NULL)
        ptr = al->n[ptr].next;
    int rec = al_get_insert_index(al);
    if (rec == AL_NULL) return;
    al->n[ptr].next = al->current = rec;
    al->n[rec].data = data;
    al->n[rec].next = AL_NULL;
    al->n[rec].dnext = AL_NULL;
    al->no++;
}

int al_search(AList *al, int data) {
    int ptr = al->head;
    while (ptr != AL_NULL) {
        if (al->n[ptr].data == data) {
            al->current = ptr;
            return ptr;
        }
        ptr = al->n[ptr].next;
    }
    return AL_NULL;
}

void al_remove_first(AList *al) {
    if (al->head == AL_NULL) return;
    int ptr = al->head;
    al->head = al->current = al->n[ptr].next;
    al->n[ptr].dnext = al->deleted;
    al->deleted = ptr;
    al->no--;
}
